### PR TITLE
Feat: Add replay command

### PR DESCRIPTION
**Pull request by Jebble**
This PR will

Add a command `eventstore:replay <stream> <event>` that will replay the event from eventstore.

Supports a range of events by sending start-end events like
`php artisan eventstore:replay accounts-v13 390-400`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -307,6 +307,16 @@ class QuoteStartedTest extends TestCase
...
 
+## Usage - Replaying Events
+
+You can replay events by using the replay command
+
+````
+php artisan eventstore:replay <stream> <event>
+````
+
+`<event>` can be a single event number of a range like `390-396`
```
Comment (morrislaptop) [inline README.md]:
```suggestion
`<event>` can be a single event number or a range like `390-396`
```

File: src/Console/Commands/EventStoreReplay.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        $data->setMetadata(json_encode($this->safeGetMetadata($event)));
+        $data->setCreatedEpoch(strtotime($event['updated']) * 1000);
+
+        return new JsonEventRecord($data);
+    }
+
+    private function safeGetMetadata(array $event)
+    {
+        return $event['content']['metadata'] === '' ? [] : $event['content']['metadata'];
+    }
```
Comment (morrislaptop) [inline src/Console/Commands/EventStoreReplay.php]: This is exact same code as the `EventStoreWorkerThread` should we refactor to trait? Service class? Leave as is?

**Review by morrislaptop — APPROVED** [review]

File: src/Console/Commands/EventStoreReplay.php (hunk 2 above)
Comment (eithed) [reply src/Console/Commands/EventStoreReplay.php]: Yes. Both trait and service class is fine with me. No

File: src/Console/Commands/EventStoreReplay.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    public function handle()
+    {
+        $stream = $this->argument('stream');
+        $events = explode('-', $this->argument('events'));
+
+        if (count($events) > 1) {
+            $events = range($events[0], $events[1]);
+        }
+
```
Comment (eithed) [inline src/Console/Commands/EventStoreReplay.php:50]: 🧗 It doesn't handle case we I pass `foo-bar`. Please throw exception if the range is invalid - `filter_validate_int` should be helpful.

Separate issue is - when we define 0-1000000 on a stream where there are not enough events. This is not handled as well.

**Review by eithed — CHANGES_REQUESTED** [review]
Please also remove duplicate code between commands.

One should think about this command as Laravel Listeners Rerun; I can see it being beneficial, but also explosively dangerous.

**Comment by morrislaptop** [comment]
> One should think about this command as Laravel Listeners Rerun; I can see it being beneficial, but also explosively dangerous.

Only accessible via the CLI so I'm OK with this.

**Comment by eithed** [comment]
> > One should think about this command as Laravel Listeners Rerun; I can see it being beneficial, but also explosively dangerous.
> 
> Only accessible via the CLI so I'm OK with this.

Yeah - I was thinking of admin error, but we cannot do anything about this.

File: src/Console/Commands/EventStoreReplay.php (hunk 3 above)
Comment (Jebble) [reply src/Console/Commands/EventStoreReplay.php:50]: >Separate issue is - when we define 0-1000000 on a stream where there are not enough events. This is not handled as well.

Just thinking from a usage perspective, this would be called with a very specific event or range in mind. I wanted to prevent checking if the first and last event exist and then going back and retrieving the events again. It felt like clutter to me. Would you be happy with an exception in the foreach if the response statusCode !== 200?

Comment (eithed) [reply src/Console/Commands/EventStoreReplay.php:50]: Yeah - exception would be good for first event that doesn't exist. You are still checking if event exists as that's what ` $this->client->get("/streams/$stream/$event"` is doing. Simply handling it somehow when 404 will be enough

Comment (Jebble) [reply src/Console/Commands/EventStoreReplay.php:50]: Added test for 404, but Guzzle already throws an exception so it is actually handled and breaks the foreach.

**Comment by Jebble** [comment]
@eithed Added extra tests and exceptions and moved duplicate code to service class

**Comment by Jebble** [comment]
@morrislaptop It's this one! Ready to merge, what's the release process for this one?

**Comment by morrislaptop** [comment]
I think it will release automatically once merged! Otherwise just tag a feature version then composer update on our prop services
